File: infrastructure/stacks/StandardStandAloneCodeBuildProject.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';

import { Construct } from 'constructs';

import * as cdk from 'aws-cdk-lib/core';
import * as aws_iam from 'aws-cdk-lib/aws-iam';
import * as aws_ec2 from 'aws-cdk-lib/aws-ec2';
import * as aws_logs from 'aws-cdk-lib/aws-logs';
import * as aws_codebuild from 'aws-cdk-lib/aws-codebuild';
import * as aws_codecommit from 'aws-cdk-lib/aws-codecommit';
import * as aws_secretsmgr from 'aws-cdk-lib/aws-secretsmanager';

import * as constants from '../constants';

//// -----------------------------------------------------------------

enum CodeBuildProjects {
    AUtomatedTesting = 'auto-testing',
    // Gateway = 'gateway',
    // CMS = 'cms',
    // Client = 'client',
}

//// -----------------------------------------------------------------

function getCodeBuildEnvVars( stk : cdk.Stack,
    props : CdkCodeBuildStackProps,
    vpc : aws_ec2.IVpc,
    whichCodeBuildProject: CodeBuildProjects,
 ): { [key: string]: aws_codebuild.BuildEnvironmentVariable } {

    const gitRepoName = `${constants.CDK_APP_NAME}-${whichCodeBuildProject}`;
    const EcrRepository = `${gitRepoName}-repo`;

    const publicSubnets = vpc.publicSubnets
        .filter(s => vpc.availabilityZones.includes(s.availabilityZone))
        .map(s => s.subnetId)
        .join(',');

    const privateSubnets = vpc.privateSubnets
        .filter(s => vpc.availabilityZones.includes(s.availabilityZone))
        .map(s => s.subnetId)
        .join(',');

    //// Load the "OVERRIDING" entries from `./.env` and EVAL the strings in them.
    const overrides = loadEnvVarsFromFile( stk, whichCodeBuildProject );

    //// ---------------------------------------
    switch (whichCodeBuildProject) {
        case CodeBuildProjects.AUtomatedTesting:
        // case CodeBuildProjects.Gateway:
        // case CodeBuildProjects.CMS:
        // case CodeBuildProjects.Client:
        return {
            APP_ENV: { value: props.envName },
            ENV: { value: props.envName }, //// used within build.sh

            DOMAIN_NAME: { value: overrides.DOMAIN_NAME ??  "??????" },
            CERTIFICATE_ARN: { value: overrides.CERTIFICATE_ARN ??  "???????" },
            ECR_REPOSITORY: { value: overrides.ECR_REPOSITORY ??  EcrRepository },

            AWS_ACCOUNT_ID: { value: stk.account },
            AWS_REGION: { value: stk.region },
            VPC_ID: { value: props.vpcId },
            AVAILABILITY_ZONES: { value: vpc.availabilityZones.join(',') },
            SECURITY_GROUP: { value: props.securityGroup },
            PUBLIC_SUBNETS: { value: publicSubnets },
            PRIVATE_SUBNETS: { value: privateSubnets },
        };
        default: return {};
    } // switch
} // getCodeBuildEnvVars()

//// -----------------------------------------------------------------

const BUILDSPEC_FILE_PATH = 'buildspec.yaml';

interface CdkCodeBuildStackProps {
    readonly envName: string;
    // readonly projectName: string;
    readonly vpcId: string;
    // readonly availabilityZones: string[];
    readonly securityGroup: string;
    // readonly BuildSpecFilepath: string;
    readonly gitBranchOrGitRef: string;
    // readonly EcrRepository: string;
}

//// -----------------------------------------------------------------

/**
 * Want to add Env-Vars to it?  Use a `.env` file at top of project.
 * In that `.env` file you can use `${this.<property>}` to reference CDK properties like account, region, etc.
 *
 * In addition to what is in `.env`, the CodeBuild-project will automatically have the following ENV-VARS:-
 *  - `AWS_ACCOUNT_ID`
 *  - `AWS_REGION`
 *  - `APP_ENV`
 *  - `VPC_ID`
 *  - `AVAILABILITY_ZONES`
 *  - `SECURITY_GROUP`
 *  - `PUBLIC_SUBNETS`
 *  - `PRIVATE_SUBNETS`
 *  - `ECR_REPOSITORY`
 */
export class CdkCodeBuildStack extends cdk.Stack {
    constructor(
        scope: Construct,
        id: string,
        props: CdkCodeBuildStackProps,
        stdCdkStkProps: cdk.StackProps,
    ) {
        super(scope, id, stdCdkStkProps);

        const vpc = aws_ec2.Vpc.fromLookup(this, 'Vpc', { vpcId: props.vpcId });


        for ( const buildspecType of Object.values(CodeBuildProjects) ) {
            const buildSpecFilepath = `config/${props.envName}-${buildspecType}-${BUILDSPEC_FILE_PATH}`;
            this.oneSingleCodeBuild( this,
                        buildspecType,
                        vpc,
                        // publicSubnets, privateSubnets,
                        buildSpecFilepath,
                        props,
                    );
        } // for loop over `CodeBuildProjects`

    } // end constructor

    //// --------------------------------------------------------------
    private oneSingleCodeBuild(
        stk: cdk.Stack,
        whichCodeBuildProject: CodeBuildProjects,
        vpc: aws_ec2.IVpc,
        // publicSubnets: string,
        // privateSubnets: string,
        buildSpecFilepath: string,
        props: CdkCodeBuildStackProps,
    ) {
        //// ------------------------------------------------------
        const gitRepoName = `${constants.CDK_APP_NAME}-${whichCodeBuildProject}`;
        const projectName = `${gitRepoName}-${props.envName}-build`;
        //// ------------------------------------------------------
        const buildSpec = this.getBuildSpec(buildSpecFilepath);
        const cbproj = new aws_codebuild.Project( stk, 'CodeBuildProject='+whichCodeBuildProject, {
            projectName: projectName,
            buildSpec: buildSpec,
            environment: {
                buildImage: aws_codebuild.LinuxBuildImage.STANDARD_7_0,
                environmentVariables: {
                    ...getCodeBuildEnvVars( stk, props, vpc, whichCodeBuildProject ),
                },
            },
            logging: {
                cloudWatch: {
                    logGroup: new aws_logs.LogGroup( stk, 'CodeBuildLogGroup-'+whichCodeBuildProject, {
                        retention: aws_logs.RetentionDays.THREE_MONTHS,
                        removalPolicy: cdk.RemovalPolicy.DESTROY,
                    }),
                },
            },
        });

        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccesstoQdrantSecret',
            actions: [ 'secretsmanager:GetSecretValue' ],
            resources: [
                `arn:aws:secretsmanager:${stk.region}:${stk.account}:secret:*`,
            ],
        }));

        // Add CloudFormation permissions
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCloudFormation',
            actions: [ 'cloudformation:*' ],
            resources: ['*'],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: "TODOAllowSSMActions",
            actions: [
                "ssm:*",
                "cloudformation:*"
            ],
            resources: ['*'],
        }));
        // cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
        //     actions: [
        //         'ec2:Describe*',
        //         'ec2:Get*',
        //         'ec2:List*',
        //     ],
        //     resources: ['*'],
        // }));

        // Add S3 permissions for CDK assets
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCDKAssetsBucket',
            actions: [ 's3:*' ],
            resources: [
                `arn:aws:s3:::cdk-*-assets-${stk.account}-${stk.region}`,
                `arn:aws:s3:::cdk-*-assets-${stk.account}-${stk.region}/*`,
            ],
        }));

        // Add permissions to assume CDK bootstrap roles
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCDKStandardRoles',
            actions: [ 'sts:AssumeRole' ],
            resources: [
                `arn:aws:iam::${stk.account}:role/cdk-*-deploy-role-${stk.account}-${stk.region}`,
                `arn:aws:iam::${stk.account}:role/cdk-*-file-publishing-role-${stk.account}-${stk.region}`,
                `arn:aws:iam::${stk.account}:role/cdk-*-image-publishing-role-${stk.account}-${stk.region}`,
            ],
        }));


        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCodeArtifactBearerToken',
            actions: ["sts:GetServiceBearerToken"],
            resources: ["*"],
            conditions: {
                "StringEquals": {
                    "sts:AWSServiceName": "codeartifact.amazonaws.com"
                }
            }
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCodeArtifactToken',
            actions: ["codeartifact:GetAuthorizationToken"],
            resources: ['*'],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCodeArtifact',
            actions: [
                    "codeartifact:List*",
                    "codeartifact:Describe*",
                    "codeartifact:Get*",
                    "codeartifact:Read*",
                    "sts:GetServiceBearerToken"
                ],
            resources: [
                    `arn:aws:codeartifact:${stk.region}:${stk.account}:domain/veridix`,
                    `arn:aws:codeartifact:${stk.region}:${stk.account}:repository/veridix/*`
                ],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToECRToken',
            actions: [
                "ecr:GetAuthorizationToken",
                "ecr:DescribeRepositories",
                "ecr:CreateRepository",
                "ecr:SetRepositoryPolicy",
                "ecr:TagResource",
            ],
            resources: ['*'],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToECRepoImages',
            actions: [
                "ecr:GetAuthorizationToken",
                "ecr:DescribeRepositories",
                "ecr:CreateRepository",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "ecr:DescribeImages",
                "ecr:PutImage",
                "ecr:InitiateLayerUpload",
                "ecr:UploadLayerPart",
                "ecr:CompleteLayerUpload",
            ],
            resources: [ // '*'
                    `arn:aws:ecr:${stk.region}:${stk.account}:repository/*`,
                    `arn:aws:ecr:${stk.region}:${stk.account}:repository/*`,
            ],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToDNS',
            actions: [
                "route53:Get*",
                "route53:List*"
            ],
            resources: ['*'],
        }));
        cbproj.addToRolePolicy(new aws_iam.PolicyStatement({
            sid: 'AccessToCloudFront',
            actions: [
                "cloudfront:Get*",
                "cloudfront:List*",
                // "cloudfront:Create*",
                // "cloudfront:Update*",
                // "cloudfront:Delete*",
                // "cloudfront:Tag*",
            ],
            resources: ['*'],
        }));

    } // oneSingleCodeBuild()

    //// --------------------------------------------------------------
    private getBuildSpec(buildSpecFilepath: string): aws_codebuild.BuildSpec {
        const fullPath = path.join(__dirname, '..', buildSpecFilepath);
        const stats = fs.statSync(fullPath);
        const MAX_INLINE_SIZE = 25600; // CodeBuild inline buildspec limit in bytes

        if (stats.size <= MAX_INLINE_SIZE) {
            const content = fs.readFileSync(fullPath, 'utf-8');
            return aws_codebuild.BuildSpec.fromObject(JSON.parse(JSON.stringify(require('js-yaml').load(content))));
        }
        return aws_codebuild.BuildSpec.fromAsset(buildSpecFilepath);
    }
} // class CdkCodeBuildStack.

//// ==================================================

/**
 * Some values like CERT-ARN, which are sensitive information, can override the above HARDCODED CodeBuild-Env-Variables (in constructor, above).
 *
 * For that create a `./env` file.
 *
 * @returns KV-pairs from `./.env` file.
 */
function loadEnvVarsFromFile(
    stk : cdk.Stack,
    whichCodeBuildProject: CodeBuildProjects,
): Record<string, string> {
// ): Record<string, aws_codebuild.BuildEnvironmentVariable> {

    const envPath = path.join(__dirname, '..', '.env-'+whichCodeBuildProject );
    const envContent = fs.readFileSync(envPath, 'utf-8');
    const envVars: Record<string, string> = {};
    // const envVars: Record<string, aws_codebuild.BuildEnvironmentVariable> = {};

    envContent.split('\n').forEach(line => {
        const trimmed = line.trim();
        if (trimmed && !trimmed.startsWith('#')) {
            const [key, ...valueParts] = trimmed.split('=');
            if (key && valueParts.length > 0) {
                let value : string = valueParts.join('=').trim().replace(/^["']|["']$/g, '');
                value = value.replace(/\$\{stk\.partition\}/g, stk.partition);
                value = value.replace(/\$\{stk\.region\}/g, stk.region);
                value = value.replace(/\$\{stk\.account\}/g, stk.account);
                value = value.replace(/\$\{this\.partition\}/g, stk.partition);
                value = value.replace(/\$\{this\.region\}/g, stk.region);
                value = value.replace(/\$\{this\.account\}/g, stk.account);
                envVars[key.trim()] = value;
                // envVars[key.trim()] = { value };
            }
        }
    });

    return envVars;
};

File: infrastructure/stacks/standard_standalone_codebuild_project.py
"""
CDK Stack for creating standalone CodeBuild projects for automated testing.

This module creates CodeBuild projects with:
- Standard build environments
- IAM permissions for CDK deployments, ECR, CodeArtifact, etc.
- CloudWatch logging
- Environment variable configuration from .env files
"""

import os
import re
from enum import Enum
from typing import Dict, Optional, List
from pathlib import Path
import yaml

from constructs import Construct
from aws_cdk import (
    Stack,
    StackProps,
    RemovalPolicy,
    Duration,
    aws_iam,
    aws_ec2,
    aws_logs,
    aws_codebuild,
)

### ======================================================================================================

# CDK_APP_NAME constant - matches project structure
CDK_APP_NAME = "nci-CTRI-ARTI-webtools"

BUILDSPEC_FILE_PATH = "buildspec.yaml"
BUILDSPEC_MAX_INLINE_SIZE = 25600  # CodeBuild inline buildspec limit in bytes

class CodeBuildProjects(Enum):
    """Enum for different CodeBuild project types."""
    AUTOMATED_TESTING = "auto-testing"
    # GATEWAY = "gateway"
    # CMS = "cms"
    # CLIENT = "client"

### ===================================

class CdkCodeBuildStackProps:
    """Properties for CdkCodeBuildStack."""

    def __init__(
        self,
        tier: str,
        aws_env: str,
        vpc_id: str,
        subnets: list[str],
        dmz_subnets: Optional[list[str]] = None,
        webapp_subnets: Optional[list[str]] = None,
        db_subnets: Optional[list[str]] = None,
        security_group: Optional[str] = None,
    ):
        self.tier = tier
        self.aws_env = aws_env
        self.vpc_id = vpc_id
        self.subnets = subnets
        ### optional properties
        self.dmz_subnets = dmz_subnets
        self.webapp_subnets = webapp_subnets
        self.db_subnets = db_subnets
        self.security_group = security_group

### ======================================================================================================


def get_codebuild_env_vars(
    stk: Stack,
    props: CdkCodeBuildStackProps,
    vpc: aws_ec2.IVpc,
    which_codebuild_project: CodeBuildProjects,
) -> Dict[str, aws_codebuild.BuildEnvironmentVariable]:
    """
    Get environment variables for CodeBuild project.

    Args:
        stk: CDK Stack instance
        props: Stack properties
        vpc: VPC for the CodeBuild project
        which_codebuild_project: Type of CodeBuild project

    Returns:
        Dictionary of environment variables for CodeBuild
    """

    git_repo_name = f"{CDK_APP_NAME}-{which_codebuild_project.value}"
    ecr_repository = f"{git_repo_name}-repo"

    ### Since CloudOne aws-acct has -NO- concept of public-subnet !!
    # public_subnets = ",".join([
    #     subnet.subnet_id
    #     for subnet in vpc.public_subnets   ### Public !
    #     if subnet.availability_zone in vpc.availability_zones
    # ])

    ### Since CloudOne aws-acct has 3 types of "egress" subnets.. we have to explicitly specify them!
    # private_subnets = ",".join([
    #     subnet.subnet_id
    #     for subnet in vpc.private_subnets   ### Private !!
    #     if subnet.availability_zone in vpc.availability_zones
    # ])

    # Load overriding entries from .env file
    overrides = load_env_vars_from_file(stk, which_codebuild_project)

    # Build environment variables based on project type
    if which_codebuild_project == CodeBuildProjects.AUTOMATED_TESTING:
        env_vars = {
            "TIER": aws_codebuild.BuildEnvironmentVariable(value=props.tier),
            "AWS_ENV": aws_codebuild.BuildEnvironmentVariable(value=props.aws_env),
            "GIT_BRANCH": aws_codebuild.BuildEnvironmentVariable( value=overrides.get("GIT_BRANCH", "main") ),
            "FrontendStkName": aws_codebuild.BuildEnvironmentVariable(
                value=overrides.get("FrontendStkName", f"{CDK_APP_NAME}-{props.tier}-frontend")
            ),
            "AppUrl": aws_codebuild.BuildEnvironmentVariable(
                value=overrides.get("AppUrl", overrides.get("DOMAIN_NAME", "https://example.com"))
            ),

            "AWS_ACCOUNT_ID": aws_codebuild.BuildEnvironmentVariable(value=stk.account),
            "AWS_REGION": aws_codebuild.BuildEnvironmentVariable(value=stk.region),
            "VPC_ID": aws_codebuild.BuildEnvironmentVariable(value=props.vpc_id),
            "AVAILABILITY_ZONES": aws_codebuild.BuildEnvironmentVariable( value=",".join(vpc.availability_zones) ),
            "SECURITY_GROUP": aws_codebuild.BuildEnvironmentVariable(value=props.security_group),
            "SUBNETS": aws_codebuild.BuildEnvironmentVariable(value=",".join(props.subnets)),
            "DMZ_SUBNETS": aws_codebuild.BuildEnvironmentVariable(value=",".join(props.dmz_subnets) if props.dmz_subnets else ",".join(props.subnets)),
            "WEBAPP_SUBNETS": aws_codebuild.BuildEnvironmentVariable(value=",".join(props.webapp_subnets) if props.webapp_subnets else ",".join(props.subnets)),
            "DB_SUBNETS": aws_codebuild.BuildEnvironmentVariable(value=",".join(props.db_subnets) if props.db_subnets else ",".join(props.subnets)),
        }

        # Add Secrets Manager ARNs as plain env vars - these ARNs are used in buildspec's secrets-manager section
        # This way the actual secret values are only accessible via secrets-manager syntax, not as regular env vars
        if "AUTO_TESTING_USER_ID_SECRET_ARN" in overrides:
            env_vars["AUTO_TESTING_USER_ID_SECRET_ARN"] = aws_codebuild.BuildEnvironmentVariable(
                value=overrides["AUTO_TESTING_USER_ID_SECRET_ARN"]
            )

        if "AUTO_TESTING_USER_PASSWORD_SECRET_ARN" in overrides:
            env_vars["AUTO_TESTING_USER_PASSWORD_SECRET_ARN"] = aws_codebuild.BuildEnvironmentVariable(
                value=overrides["AUTO_TESTING_USER_PASSWORD_SECRET_ARN"]
            )

        return env_vars

    return {}



### ======================================================================================================


class CdkCodeBuildStack(Stack):
    """
    CDK Stack for creating CodeBuild projects.

    Want to add Env-Vars to it? Use a `.env` file at top of project.
    In that `.env` file you can use `${this.<property>}` to reference CDK properties like account, region, etc.

    In addition to what is in `.env`, the CodeBuild-project will automatically have the following ENV-VARS:
    - AWS_ACCOUNT_ID
    - AWS_REGION
    - APP_ENV
    - VPC_ID
    - AVAILABILITY_ZONES
    - SECURITY_GROUP
    - PUBLIC_SUBNETS
    - PRIVATE_SUBNETS
    - ECR_REPOSITORY
    """

    def __init__(
        self,
        scope: Construct,
        id: str,
        props: CdkCodeBuildStackProps,
        **kvargs,
    ) -> None:
        super().__init__(scope, id, **kvargs)

        vpc = aws_ec2.Vpc.from_lookup(self, "Vpc", vpc_id=props.vpc_id)

        # Create one CodeBuild project for each project type
        for buildspec_type in CodeBuildProjects:
            buildspec_filepath = f"config/{props.tier}-{buildspec_type.value}-{BUILDSPEC_FILE_PATH}"
            self._create_single_codebuild(
                buildspec_type,
                vpc,
                buildspec_filepath,
                props,
            )

    ### -------------------------------------------------------------------------

    def _create_single_codebuild(
        self,
        which_codebuild_project: CodeBuildProjects,
        vpc: aws_ec2.IVpc,
        buildspec_filepath: str,
        props: CdkCodeBuildStackProps,
    ) -> None:
        """Create a single CodeBuild project with standard permissions."""

        git_repo_name = f"{CDK_APP_NAME}-{which_codebuild_project.value}"
        project_name = f"{git_repo_name}-{props.tier}-build"

        ### Get CodeBuild's buildspec from a file.
        full_path = Path(__file__).parent.parent / buildspec_filepath # Path relative to infrastructure directory
        if full_path.exists():
            stats = full_path.stat()
            if stats.st_size <= BUILDSPEC_MAX_INLINE_SIZE:
                with open(full_path, "r") as f:
                    content = yaml.safe_load(f)
                build_spec = aws_codebuild.BuildSpec.from_object(content)
            else:
                build_spec = aws_codebuild.BuildSpec.from_asset(buildspec_filepath)
        else:
            raise FileNotFoundError(f"Buildspec file not found: {full_path}")

        ### Create CodeBuild project
        cbproj = aws_codebuild.Project(
            self,
            f"CodeBuildProject={which_codebuild_project.value}",
            project_name=project_name,
            build_spec=build_spec,
            environment=aws_codebuild.BuildEnvironment(
                build_image=aws_codebuild.LinuxBuildImage.STANDARD_7_0,
                environment_variables=get_codebuild_env_vars(
                    self, props, vpc, which_codebuild_project
                ),
            ),
            logging=aws_codebuild.LoggingOptions(
                cloud_watch=aws_codebuild.CloudWatchLoggingOptions(
                    log_group=aws_logs.LogGroup(
                        self,
                        f"CodeBuildLogGroup-{which_codebuild_project.value}",
                        retention=aws_logs.RetentionDays.THREE_MONTHS,
                        removal_policy=RemovalPolicy.DESTROY,
                    ),
                ),
            ),
        )

        ### Add IAM permissions
        self._add_iam_permissions(cbproj)

    ### -------------------------------------------------------------------------

    def _add_iam_permissions(self, cbproj: aws_codebuild.Project) -> None:
        """Add standard IAM permissions to CodeBuild project role."""

        ### Secrets Manager access
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccesstoQdrantSecret",
                actions=["secretsmanager:GetSecretValue"],
                resources=[
                    f"arn:aws:secretsmanager:{self.region}:{self.account}:secret:*",
                ],
            )
        )

        ### CloudFormation permissions
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCloudFormation",
                actions=["cloudformation:*"],
                resources=["*"],
            )
        )

        ### SSM and CloudFormation
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="TODOAllowSSMActions",
                actions=["ssm:*", "cloudformation:*"],
                resources=["*"],
            )
        )

        ### S3 permissions for CDK assets
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCDKAssetsBucket",
                actions=["s3:*"],
                resources=[
                    f"arn:aws:s3:::cdk-*-assets-{self.account}-{self.region}",
                    f"arn:aws:s3:::cdk-*-assets-{self.account}-{self.region}/*",
                ],
            )
        )

        ### Permissions to assume CDK bootstrap roles
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCDKStandardRoles",
                actions=["sts:AssumeRole"],
                resources=[
                    f"arn:aws:iam::{self.account}:role/cdk-*-deploy-role-{self.account}-{self.region}",
                    f"arn:aws:iam::{self.account}:role/cdk-*-file-publishing-role-{self.account}-{self.region}",
                    f"arn:aws:iam::{self.account}:role/cdk-*-image-publishing-role-{self.account}-{self.region}",
                ],
            )
        )

        ### CodeArtifact bearer token
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCodeArtifactBearerToken",
                actions=["sts:GetServiceBearerToken"],
                resources=["*"],
                conditions={
                    "StringEquals": {"sts:AWSServiceName": "codeartifact.amazonaws.com"}
                },
            )
        )

        ### CodeArtifact authorization token
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCodeArtifactToken",
                actions=["codeartifact:GetAuthorizationToken"],
                resources=["*"],
            )
        )

        ### CodeArtifact access
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCodeArtifact",
                actions=[
                    "codeartifact:List*",
                    "codeartifact:Describe*",
                    "codeartifact:Get*",
                    "codeartifact:Read*",
                    "sts:GetServiceBearerToken",
                ],
                resources=[
                    f"arn:aws:codeartifact:{self.region}:{self.account}:domain/veridix",
                    f"arn:aws:codeartifact:{self.region}:{self.account}:repository/veridix/*",
                ],
            )
        )

        ### ECR token access
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToECRToken",
                actions=[
                    "ecr:GetAuthorizationToken",
                    "ecr:DescribeRepositories",
                    "ecr:CreateRepository",
                    "ecr:SetRepositoryPolicy",
                    "ecr:TagResource",
                ],
                resources=["*"],
            )
        )

        ### ECR repository images
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToECRepoImages",
                actions=[
                    "ecr:GetAuthorizationToken",
                    "ecr:DescribeRepositories",
                    "ecr:CreateRepository",
                    "ecr:BatchCheckLayerAvailability",
                    "ecr:GetDownloadUrlForLayer",
                    "ecr:BatchGetImage",
                    "ecr:DescribeImages",
                    "ecr:PutImage",
                    "ecr:InitiateLayerUpload",
                    "ecr:UploadLayerPart",
                    "ecr:CompleteLayerUpload",
                ],
                resources=[
                    f"arn:aws:ecr:{self.region}:{self.account}:repository/*",
                ],
            )
        )

        ### Route53 DNS access
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToDNS",
                actions=["route53:Get*", "route53:List*"],
                resources=["*"],
            )
        )

        ### CloudFront access
        cbproj.add_to_role_policy(
            aws_iam.PolicyStatement(
                sid="AccessToCloudFront",
                actions=[
                    "cloudfront:Get*",
                    "cloudfront:List*",
                    # "cloudfront:Create*",
                    # "cloudfront:Update*",
                    # "cloudfront:Delete*",
                    # "cloudfront:Tag*",
                ],
                resources=["*"],
            )
        )


### ======================================================================================================


def load_env_vars_from_file(
    stk: Stack,
    which_codebuild_project: CodeBuildProjects,
) -> Dict[str, str]:
    """
    UTILITY Function !!
    To help test this CDK-code locally on GFE.  Simulates env-vars specified in GH-Actions.

    Loads environment variables from `.env***` files.
    NOTE: Each CodeBuild-project will need have its --OWN--  `./.env-<project-type>` file !!

    Some values can be sensitive information, can override the hardcoded/placeholder CodeBuild env variables.

    Args:
        stk: CDK Stack instance
        which_codebuild_project: Type of CodeBuild project

    Returns:
        Dictionary of environment variable overrides
    """
    env_path = Path(__file__).parent.parent / f".env-{which_codebuild_project.value}"
    env_vars: Dict[str, str] = {}

    if not env_path.exists():
        return env_vars

    with open(env_path, "r") as f:
        env_content = f.read()

    for line in env_content.split("\n"):
        trimmed = line.strip()
        if trimmed and not trimmed.startswith("#"):
            parts = trimmed.split("=", 1)
            if len(parts) == 2:
                key = parts[0].strip()
                value = parts[1].strip().strip('"').strip("'")

                # Replace CDK property references
                value = re.sub(r'\$\{stk\.partition\}', stk.partition, value)
                value = re.sub(r'\$\{stk\.region\}', stk.region, value)
                value = re.sub(r'\$\{stk\.account\}', stk.account, value)
                value = re.sub(r'\$\{this\.partition\}', stk.partition, value)
                value = re.sub(r'\$\{this\.region\}', stk.region, value)
                value = re.sub(r'\$\{this\.account\}', stk.account, value)

                env_vars[key] = value

    return env_vars


### ======================================================================================================
### EoF
